### PR TITLE
node numbering added for node labels on the circosplot

### PR DESCRIPTION
**Pull request by ThijsHrm**
Hey Eric, I saw an opportunity to add my code to the nxviz package. It's my first open source contribution so I hope I did everything right. If not please let me know so I can learn from it and fix it! Also I fixed a minor bug related to group labeling on line 179 which is unrelated but can cause errors otherwise (see changes below). Is that a okay or should I do that seperately? Kind regards, Thijs

### CODE REVIEW

**Comment by ThijsHrm** [comment]
Fixing

**Comment by ThijsHrm** [comment]
Pull linked to #75

**Comment by ericmjl** [comment]
@ThijsHrm thank you for pushing this! I will now embark on a code review with line comments.

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -84,7 +84,7 @@ Ready to contribute? Here's how to set up `nxviz` for local development.
 
    To get flake8 and tox, just pip install them into your virtualenv.
 
-6. Commit your changes and push your branch to GitHub::
+6. Commit your changes and push your branch to GitHub:
```
Comment (ericmjl) [inline CONTRIBUTING.rst]: The double colon (`::`) is Sphinx syntax to indicate that the coming block is to be formatted like code. It's a minor thing, so I'll change this back on my own, no need to worry about this :smile:.

File: AUTHORS.rst [hunk 2]
```diff
@@ -12,3 +12,4 @@ Contributors
 
 * Nora Kassner <nora.kassner@t-online.de>
 * Christian Diener <mail (at) cdiener.com>
+* Thijs Hermans <hermansthijs91@gmail.com>
```
Comment (ericmjl) [inline AUTHORS.rst]: WOOHOO!! :smile:

File: nxviz/plots.py [hunk 3]
```diff
@@ -338,10 +340,18 @@ def __init__(self, graph, **kwargs):
...
 
-        :param rotate_labels: Whether to rotate node labels.
-        :type node_color: `bool`
+        :param node_layout: which/whether (a) node layout is used,
+            either 'rotation', 'numbers' or None
+        :type node_layout: `string`
         """
-        self.rotate = kwargs.pop("rotate_labels", False)
+
+        # Store node layout
```
Comment (ericmjl) [inline nxviz/plots.py]: If possible, could you elaborate upon the comment line for documentation purposes? I'm preparing for the possibility of forgetting what this block of code was all about 😅. Here's my attempt as an example:

```
# Specify what the node label layout will be. 
# Only three are provided: None, 'rotation', and 'numbers'. 
# Rotation: labels are drawn along the radial axis with appropriate rotation and alignment (e.g. a label at the 2 o'clock position will be rotated +30º from the horizontal and left-aligned, while a label at the 10 o'clock position will be rotated -30º from the horizontal and right-aligned).
# Numbers: labels are rendered along the circumference with appropriate alignment, but are not rotated (i.e. they are either horizontal left-aligned or horizontal right-aligned).
```

File: nxviz/plots.py [hunk 4]
```diff
@@ -338,10 +340,18 @@ def __init__(self, graph, **kwargs):
...
+        specified_layout = kwargs.pop("node_layout", None)
```
Comment (ericmjl) [inline nxviz/plots.py]: On a related note, this is not technically a "node layout", but a "node label layout". This is the first place I'm noticing the use of `node_layout`, for clarity I'd like to rename this to `node_label_layout` please. I will mark out the other places where I'd like this changed.

File: nxviz/plots.py [hunk 5]
```diff
@@ -338,10 +340,18 @@ def __init__(self, graph, **kwargs):
...
+        valid_node_layouts = (None, 'rotation', 'numbers')
```
Comment (ericmjl) [inline nxviz/plots.py]: Please change `valid_node_layouts` to `valid_node_label_layouts`.

File: nxviz/plots.py [hunk 6]
```diff
@@ -338,10 +340,18 @@ def __init__(self, graph, **kwargs):
...
+        assert specified_layout in valid_node_layouts
```
Comment (ericmjl) [inline nxviz/plots.py]: Please change `valid_node_layouts` to `valid_node_label_layouts`. I will leave it to you to decide whether or not to change `specified_layout` to `specified_node_layout` or not.

File: nxviz/plots.py [hunk 7]
```diff
@@ -338,10 +340,18 @@ def __init__(self, graph, **kwargs):
...
+        self.node_layout = specified_layout
```
Comment (ericmjl) [inline nxviz/plots.py]: Please change `self.node_layout` to `self.node_label_layout`.

File: nxviz/plots.py [hunk 8]
```diff
@@ -425,32 +435,36 @@ def compute_node_label_positions(self):
...
+            # multiplication factor 1.02 moved below
+            radius = self.plot_radius + self.nodeprops['radius']
 
-            if self.rotate:
-                va = "center"
+            # Coordinates of text inside nodes
+            radius_adjustment = 0.96 if self.node_layout == 'numbers' else 1.02
+            x, y = get_cartesian(r=radius * radius_adjustment, theta=theta)
+
+            # For numbered nodes
```
Comment (ericmjl) [inline nxviz/plots.py]: A future coder reading this code might need more visual clues to understand why `# For numbered nodes` exists as its own line. Perhaps add a few `-` characters before and after to signify that it's like a "heading"?

File: nxviz/plots.py [hunk 9]
```diff
@@ -425,32 +435,36 @@ def compute_node_label_positions(self):
...
+            tx += self.nz_sign(x)
+
+            # Node description y numerator
+            numer = \
+                radius * (theta % (self.nz_sign(y) * self.nz_sign(x) * np.pi))
+            # Node description y denominator
+            denom = (self.nz_sign(x) * np.pi)
+            ty = 2 * (numer / denom)
+
+            # For rotated nodes
```
Comment (ericmjl) [inline nxviz/plots.py]: As with `# For numbered nodes`, same point. Perhaps add a few `-` characters before and after to signify that it's like a "heading"?

File: nxviz/plots.py [hunk 10]
```diff
@@ -459,15 +473,68 @@ def compute_node_label_positions(self):
...
-            xs.append(x)
-            ys.append(y)
-            has.append(ha)
-            vas.append(va)
-            rotations.append(rot)
+            # Store values
+            self._store_node_label_meta(x, y, tx, ty, rot)
+
+    @staticmethod
+    def nz_sign(xy):
```
Comment (ericmjl) [inline nxviz/plots.py]: Please add a docstring that describes what this function is intended to do. Also, the naming is a bit cryptic: what does `nz` in `nz_sign` stand for?

File: nxviz/plots.py [hunk 11]
```diff
@@ -459,15 +473,68 @@ def compute_node_label_positions(self):
...
+        return -1 if xy < 0 else 1
+
+    def _init_node_label_meta(self):
```
Comment (ericmjl) [inline nxviz/plots.py]: I like the extra checks that this method provides. Have you ran the tests and verified that this change doesn't break anything, and ran a few of the examples to visually check as well?

File: nxviz/plots.py [hunk 12]
```diff
@@ -486,17 +553,39 @@ def draw_nodes(self):
...
                 label_x = self.node_label_coords['x'][i]
                 label_y = self.node_label_coords['y'][i]
+                label_tx = self.node_label_coords['tx'][i]
+                label_ty = self.node_label_coords['ty'][i]
                 label_ha = self.node_label_aligns['has'][i]
                 label_va = self.node_label_aligns['vas'][i]
-                rot = 0
-                if self.rotate:
+
+                # Node rotation layout
```
Comment (ericmjl) [inline nxviz/plots.py]: Please change this to `# Node label rotation layout`.

File: nxviz/plots.py [hunk 13]
```diff
@@ -486,17 +553,39 @@ def draw_nodes(self):
...
-                             ha=label_ha, va=label_va, rotation=rot,
-                             rotation_mode="anchor",
-                             color=self.node_label_color[i], fontsize=10)
+                    self.ax.text(s=node,
+                                 x=label_x, y=label_y,
+                                 ha=label_ha, va=label_va, rotation=rot,
+                                 rotation_mode="anchor",
+                                 color=self.node_label_color[i], fontsize=10)
+
+                # Node numbering layout
```
Comment (ericmjl) [inline nxviz/plots.py]: Please change this to `# Node label numbering layout`.

File: nxviz/plots.py [hunk 14]
```diff
@@ -486,17 +553,39 @@ def draw_nodes(self):
...
+                    # Draw descriptions for labels
+                    desc = '%s - %s' % ((i, node) if (x > 0) else (node, i))
+                    self.ax.text(s=desc, x=label_tx,
+                                 y=label_ty, ha=label_ha, va=label_va)
+
+                    # Add numbers to nodes
+                    self.ax.text(s=i, x=label_x, y=label_y,
+                                 ha='center', va='center')
+
+                # Standard node layout
```
Comment (ericmjl) [inline nxviz/plots.py]: Please change this to `# Standard node label layout`.

File: nxviz/plots.py [hunk 15]
```diff
@@ -425,32 +435,36 @@ def compute_node_label_positions(self):
...
 
-            if self.rotate:
-                va = "center"
+            # Coordinates of text inside nodes
+            radius_adjustment = 0.96 if self.node_layout == 'numbers' else 1.02
+            x, y = get_cartesian(r=radius * radius_adjustment, theta=theta)
+
+            # For numbered nodes
+
+            # Node description x
```
Comment (ericmjl) [inline nxviz/plots.py]: Please change this to '# Node label x-axis coordinate`. It is a bit more descriptive this way. Also, please document that the variables that are `tx` and `ty` refer to the x- and y-axis coordinates of the text (hence, the tx and ty variable names). (Again, simply prioritizing clarity!)

File: nxviz/plots.py [hunk 16]
```diff
@@ -425,32 +435,36 @@ def compute_node_label_positions(self):
...
+            tx, _ = get_cartesian(r=radius, theta=theta)
+            # Create the quasi-circular positioning on the x axis
+            tx *= 1 - np.log(np.cos(theta) * self.nz_sign(np.cos(theta)))
+            # Move each node a little further away from the circos
+            tx += self.nz_sign(x)
+
+            # Node description y numerator
```
Comment (ericmjl) [inline nxviz/plots.py]: Please change this to `# Node label y-axis coordinate`.

File: nxviz/plots.py [hunk 17]
```diff
@@ -338,10 +340,18 @@ def __init__(self, graph, **kwargs):
         Accepts the following additional arguments apart from the ones in
         `BasePlot`:
 
-        :param rotate_labels: Whether to rotate node labels.
-        :type node_color: `bool`
+        :param node_layout: which/whether (a) node layout is used,
```
Comment (ericmjl) [inline nxviz/plots.py]: Please change this to `:param node_label_layout:`.

**Comment by ericmjl** [comment]
@ThijsHrm please also do the following:

- [ ] Add a minimal example of this new node label layout to the `examples/` directory. A reasonably-sized graph, with dummy labels ('my label 1', 'my label 2', etc.), is good enough. 
- [ ] Check that the documentation is built correctly. This one involves a medium-sized change to the CircosPlot API, so I'd like to make sure that it's reflected appropriately on the docs. (Building the docs is not a trivially simple task, and so if it helps, I'm happy to push up instructions.)

**Comment by ericmjl** [comment]
Ping @ThijsHrm, just hoping to keep this PR alive! Do let me know how I can help.

**Comment by ThijsHrm** [comment]
Hey Eric, thank you very much for your feedback and enthousiasm! It's very helpful. Appologies for the slow response, this was due to a more urgent deadline. I will get back to work before the end of the week :)

**Comment by ericmjl** [comment]
Totally understand, @ThijsHrm! :smile: Thanks for the heads-up, I'll loop back a bit later then.

**Comment by ThijsHrm** [comment]
Alright, requested changes were made, and an example was added, as shown in these pictures:

figure_3
figure_1
figure_2

As for Sphinx, yes a standard procedure would be nice. I thought I would figure it out but I'm getting nervous I'll mess something up. Also I'm currently trying to figure out why my push failed on geometry.py test in travis. It doesn't show the exact reason why the test_node_theta assertion fails and the provided seed does not reproduce the error locally on my laptop.

**Comment by ericmjl** [comment]
Funny, the seed does reproduce on my computer. Might be a linux thing? 

Regardless, I reran without the seed, and all tests passed. The problem induced by that random seed might just be a random issue, which I'm ok letting go of.

**Comment by ericmjl** [comment]
To build the docs, make sure you have all the packages as specified in `environment.yml` installed.

`cd` into the `docs/` directory, then run `make html`. The docs should show up in a subdirectory called `_build/html`. You can then open `index.html`, which is in that directory. 

On a Mac, after running `make html`, you should be able to do `open _build/html/index.html` directly. I'll have these instructions pushed into the CONTRIBUTING.rst docs page.

**Comment by ThijsHrm** [comment]
Alright, I also just passed the tests without any changes that were relevant (only comments). Still checking the docs and then the push will be complete.

**Comment by ericmjl** [comment]
I think everything looks good! There is one little change which I'm happy to tackle on my own, that is to move the CircosPlot `__init__()` docstring that documents additional arguments into the CircosPlot docstring, where it can then be shown to end-users in the docs page. 

I'm ready to merge, just let me know if you're ok too!

**Comment by ThijsHrm** [comment]
Alright I just went over the docstrings and fixed the issues myself (see new push). Now I'm done,

**Comment by ericmjl** [comment]
That’s very kind of you! Ok, merging in now.

**Comment by ThijsHrm** [comment]
@ericmjl I accidently left a print statement in line 581 of plots.py, my apologies!

**Comment by ericmjl** [comment]
All good! Thanks for the catch. I will take it out and push to master.

**Pull request merged**